### ISSUE TITLE
Ideas for new features

### ISSUE THREAD

**Issue by PaulMaly** [deployd/deployd#482]
Hello everyone! 

Deployd is a greatest and simplest tool to build backend for mobile and rich web apps. Thanks a lot to all guys, who created it and who develop it now. Recently, Deployd was accelerated in the development of and seems that we have prospect that some day it will become a 1.0 version. Awesome!

I know maybe this topic would be better in Deployd Google Community, but there still not enough activity, so I decided to post it here. Below, you can see my modest list of that I would like to see in the closest releases.

# Properties section:
- new property type to store date and time (I prefer ISO 8601 format).

# Data section:
- make object type properties editable from dashboard as separated fields (maybe in some modal dialog).

# Events section:
- add more functions to rule HTTP responses (maybe like header(), cookie() etc. ), that it was possible, for example, to set custom HTTP header to response;
- add more events, like onHead, onOptions, onQuery etc, I think it can be useful for many cases.
- maybe re-name Events to Hooks? I'm not sure, but seems it more like a Hook or not?

# Dpd library:
- for me, it will be more convenient to use Node-style (error argument first) for callbacks.

And just for your information, maybe it makes sense to look on RMR Architecture (http://www.peej.co.uk/articles/rmr-architecture.html). It's not so popular like MVC, but I think more suitable for Deployd. RMR can be useful to describe Deployd  well.

I'll be very glad for any comments from community and developers. Thanks!

**Comment by ericfong** [comment]
You can 
```
dpd.users.get({name:'Eric'}, function(user, err){
  // I am eric
  // Or you means we should swap user and err ?
});
```

**Comment by ericfong** [comment]
For RMR, I found https://github.com/flatiron/resourceful
Similar to dpd interface, not sure can reuse or not.

**Comment by PaulMaly** [comment]
> // Or you means we should swap user and err ?
Yes, NodeJS-style callbacks. In NodeJS in most cases, you will see the following signature for callbacks:
    function (err, result)

> Similar to dpd interface, not sure can reuse or not.
I don't know what Resourceful library actually is and how it can be useful for Deployd, but I talk about architecture, not existing implementations.

For example, in Deployd we have "Resource", and seems partially have "Method" components of this architecture, but we completely don't have "Representation" component.

Eric, what do you think about other ideas?

**Comment by andreialecu** [comment]
The date/time could just be saved as a number with the value of `Date.now()` (or `new Date().getTime()`), not sure if we need a separate property type.

**Comment by PaulMaly** [comment]
>The date/time could just be saved as a number

I think, it's not a good practice to store date/time as a number. Doing this, you're lost timezone and it has bad visibility and it's hard to edit from dashboard.

So, yes, we can store ISO 8601 formated date/time as a string, like: '2014-08-09T18:31:42+03', but I'm talking about format control and auto saving of current time as an option of this property type (creation date of the entry, for example). After all, we also can store boolean values as a number, but Deployd gives us separated property type for it, right?

**Comment by PaulMaly** [comment]
Ps. Simple way to get ISO 8601 formated date/time with timezone and all stuff, is: new Date().toJSON() (IE9+)

**Comment by andreialecu** [comment]
Well, if you are interested in autosaving the date from the server, then timezone shouldn't be important - it would be the timezone of the server. I think you'd be wanting the UTC date there anyway.

I'm sure there are niche cases when you'd be interested in a date with actual timezone information attached, but I think for most purposes storing as UTC is safer.

ISO 8601 has issues with sorting if you use dates in multiple time zones if you don't convert them back to UTC before sorting, which could be an issue.

**Comment by PaulMaly** [comment]
Actually, timezone important in many cases. Yes, sorting by date/time property need to consider timezone too.

**Comment by andreialecu** [comment]
Sure. For the niche cases where timezone information is actually important on a per-record basis, a separate property for the timezone specifically would solve the problem. Or, you could just save it as a string explicitly in ISO 8601 format **just** for those specific cases.

99% of the time, the timezone is not important **on a per record basis** and is only a source of bugs. I'm glad that JavaScript defaults to UTC as opposed to other languages, look at any old SQL database and you'll see that UTC is hardly ever used, leading to sorting issues for things as simple as daylight savings time.

Thinking about it, we probably do **need** a datetime type in the dashboard just so users don't get confused about what the best practice should be and use some sort of string format when they probably really shouldn't.

**Comment by ericfong** [comment]
I think db should always "2015-01-12T07:35:37.190Z" in ISO format.

For larger structure, I think dpd should be really small, simple, extensible classes and util on top of koa and expressjs.

So, I like minimal core and use npm package to extend. In that way, Date formatting should not be considered in the core.

**Comment by PaulMaly** [comment]
Eric, but boolean type already considered in the core. I think it was made for better querying or something like that. So, datetime is also very important for querying, not just for the convenience.
